### PR TITLE
public changes to support GenotypeGVCFs

### PR DESCRIPTION
**Pull request by lbergelson**
These are the changes to public necessary to support my initial port of genotypeGVCFs.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProvider.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.broadinstitute.hellbender.tools.walkers.genotyper.GenotypeCalculationArgumentCollection;
+import org.broadinstitute.hellbender.utils.Utils;
+
+/**
+ * Provider that defaults to the general ploidy implementation when the preferred one does not handle the required
+ * ploidy.
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class GeneralPloidyFailOverAFCalculatorProvider extends AFCalculatorProvider {
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProvider.java]: final class

File: src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java [hunk 2]
```diff
@@ -385,11 +393,29 @@ else if ( Double.isInfinite(expected) ) // NaN == NaN => false unfortunately
...
-        } else
-            Assert.assertEquals(actual, expected, "Attribute " + key);
+            assertEqualsDoubleSmart((Double) notationCorrectedActual, (Double) notationCorrectedExpected, 1e-2, "Attribute " + key);
+        } else {
+            Assert.assertEquals(notationCorrectedActual, notationCorrectedExpected, "Attribute " + key);
+        }
+    }
+
+    /**
+     * This is some  necessary to deal with the fact that variant context attributes are deserialized from vcf as Strings
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: `some` -> `sometimes` (?)

Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: why not put this in Utils? that way would not have to create BaseTestUnitTest

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.testng.Assert;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+/**
+ * Tests {@link GeneralPloidyFailOverAFCalculatorProvider}
+ *
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public class GeneralPloidyFailOverAFCalculatorProviderUnitTest {
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java]: `extends BaseTest` as a way to track all tests

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Test(dataProvider="nonThreadSafeConstructorsData")
+    public void testNonThreadSafeConstructors(final int ploidy, final int maxAltAlleles) {
```
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java]: do we have 'thread safe' constructors?

**Comment by lbergelson** [comment]
@akiezun Do you think we should move the various variant context test routines out of BaseTest?  It seems weird that they live there.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java (hunk 4 above)
Comment (akiezun) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java]: can you rename this test and data provider? current names make no sense i think

File: src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java (hunk 2 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: Should we move these variantcontext methods as well?  It seems like they should all go somewhere else.

Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: I think so.

File: src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java [hunk 5]
```diff
@@ -403,18 +429,18 @@ public static void assertGenotypesAreEqual(final Genotype actual, final Genotype
...
         Assert.assertEquals(actual.hasPL(), expected.hasPL(), "Genotype hasPL");
-        Assert.assertEquals(actual.hasAD(), expected.hasAD(), "Genotype hasAD");
-        Assert.assertEquals(actual.hasGQ(), expected.hasGQ(), "Genotype hasGQ");
-        Assert.assertEquals(actual.hasDP(), expected.hasDP(), "Genotype hasDP");
+        Assert.assertTrue(Arrays.equals(actual.getPL(), expected.getPL()));
 
         Assert.assertEquals(actual.hasLikelihoods(), expected.hasLikelihoods(), "Genotype haslikelihoods");
         Assert.assertEquals(actual.getLikelihoodsString(), expected.getLikelihoodsString(), "Genotype getlikelihoodsString");
         Assert.assertEquals(actual.getLikelihoods(), expected.getLikelihoods(), "Genotype getLikelihoods");
-        Assert.assertTrue(Arrays.equals(actual.getPL(), expected.getPL()));
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: why is the getPL comparison removed?

File: src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java [hunk 6]
```diff
@@ -519,5 +547,42 @@ public static void assertEqualVariants(final List<VariantContext> v1, final List
         }
     }
 
+    public static void assertVariantContextsAreEqual( final VariantContext actual, final VariantContext expected, List<String> attributesToIgnore ) {
```
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: i think this one is fine to stay here because it's all asserts

**Comment by akiezun** [comment]
small edits to do - after that, LGTM. back to @lbergelson

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java (hunk 4 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java]: done

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java (hunk 3 above)
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProviderUnitTest.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java (hunk 2 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/utils/test/BaseTest.java]: @akiezun How do you feel about moving this and the various asserts in base test to a new `VariantContextTestUtils`?  It doesn't feel like this method is generally applicable, and those other methods seem more complex and specialized than should really be in BaseTest

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProvider.java (hunk 1 above)
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/genotyper/afcalc/GeneralPloidyFailOverAFCalculatorProvider.java]: done

**Pull request merged**